File: app/src/main/java/eu/kanade/tachiyomi/data/download/DownloadService.java
package eu.kanade.tachiyomi.data.download;

import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.os.IBinder;
import android.os.PowerManager;

import com.github.pwittchen.reactivenetwork.library.ReactiveNetwork;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

import javax.inject.Inject;

import eu.kanade.tachiyomi.App;
import eu.kanade.tachiyomi.R;
import eu.kanade.tachiyomi.data.preference.PreferencesHelper;
import eu.kanade.tachiyomi.event.DownloadChaptersEvent;
import eu.kanade.tachiyomi.util.ToastUtil;
import rx.Subscription;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

public class DownloadService extends Service {

    @Inject DownloadManager downloadManager;
    @Inject PreferencesHelper preferences;

    private PowerManager.WakeLock wakeLock;
    private Subscription networkChangeSubscription;
    private Subscription queueRunningSubscription;
    private boolean isRunning;

    public static void start(Context context) {
        context.startService(new Intent(context, DownloadService.class));
    }

    public static void stop(Context context) {
        context.stopService(new Intent(context, DownloadService.class));
    }

    @Override
    public void onCreate() {
        super.onCreate();
        App.get(this).getComponent().inject(this);

        createWakeLock();

        listenQueueRunningChanges();
        EventBus.getDefault().register(this);
        listenNetworkChanges();
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        return START_STICKY;
    }

    @Override
    public void onDestroy() {
        EventBus.getDefault().unregister(this);
        queueRunningSubscription.unsubscribe();
        networkChangeSubscription.unsubscribe();
        downloadManager.destroySubscriptions();
        destroyWakeLock();
        super.onDestroy();
    }

    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    @Subscribe(sticky = true, threadMode = ThreadMode.MAIN)
    public void onEvent(DownloadChaptersEvent event) {
        EventBus.getDefault().removeStickyEvent(event);
        downloadManager.onDownloadChaptersEvent(event);
    }

    private void listenNetworkChanges() {
        networkChangeSubscription = new ReactiveNetwork().enableInternetCheck()
                .observeConnectivity(getApplicationContext())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(state -> {
                    switch (state) {
                        case WIFI_CONNECTED_HAS_INTERNET:
                            // If there are no remaining downloads, destroy the service
                            if (!isRunning && !downloadManager.startDownloads()) {
                                stopSelf();
                            }
                            break;
                        case MOBILE_CONNECTED:
                            if (!preferences.downloadOnlyOverWifi()) {
                                if (!isRunning && !downloadManager.startDownloads()) {
                                    stopSelf();
                                }
                            } else if (isRunning) {
                                downloadManager.stopDownloads();
                            }
                            break;
                        default:
                            if (isRunning) {
                                downloadManager.stopDownloads();
                            }
                            break;
                    }
                }, error -> {
                    ToastUtil.showShort(this, R.string.download_queue_error);
                    stopSelf();
                });
    }

    private void listenQueueRunningChanges() {
        queueRunningSubscription = downloadManager.getRunningSubject()
                .subscribe(running -> {
                    isRunning = running;
                    if (running)
                        acquireWakeLock();
                    else
                        releaseWakeLock();
                });
    }

    private void createWakeLock() {
        wakeLock = ((PowerManager)getSystemService(POWER_SERVICE)).newWakeLock(
                PowerManager.PARTIAL_WAKE_LOCK, "DownloadService:WakeLock");
    }

    private void destroyWakeLock() {
        if (wakeLock != null && wakeLock.isHeld()) {
            wakeLock.release();
            wakeLock = null;
        }
    }

    public void acquireWakeLock() {
        if (wakeLock != null && !wakeLock.isHeld()) {
            wakeLock.acquire();
        }
    }

    public void releaseWakeLock() {
        if (wakeLock != null && wakeLock.isHeld()) {
            wakeLock.release();
        }
    }

}

File: app/src/main/java/eu/kanade/tachiyomi/data/download/DownloadService.kt
package eu.kanade.tachiyomi.data.download

import android.app.Notification
import android.app.Service
import android.content.Context
import android.content.Intent
import android.os.IBinder
import android.os.PowerManager
import androidx.annotation.StringRes
import androidx.core.content.ContextCompat
import eu.kanade.tachiyomi.R
import eu.kanade.tachiyomi.data.notification.Notifications
import eu.kanade.tachiyomi.util.system.acquireWakeLock
import eu.kanade.tachiyomi.util.system.isConnectedToWifi
import eu.kanade.tachiyomi.util.system.isOnline
import eu.kanade.tachiyomi.util.system.isServiceRunning
import eu.kanade.tachiyomi.util.system.notificationBuilder
import eu.kanade.tachiyomi.util.system.toast
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.catch
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import logcat.LogPriority
import ru.beryukhov.reactivenetwork.ReactiveNetwork
import tachiyomi.core.util.lang.withUIContext
import tachiyomi.core.util.system.logcat
import tachiyomi.domain.download.service.DownloadPreferences
import uy.kohesive.injekt.injectLazy

/**
 * This service is used to manage the downloader. The system can decide to stop the service, in
 * which case the downloader is also stopped. It's also stopped while there's no network available.
 * While the downloader is running, a wake lock will be held.
 */
class DownloadService : Service() {

    companion object {

        private val _isRunning = MutableStateFlow(false)
        val isRunning = _isRunning.asStateFlow()

        /**
         * Starts this service.
         *
         * @param context the application context.
         */
        fun start(context: Context) {
            val intent = Intent(context, DownloadService::class.java)
            ContextCompat.startForegroundService(context, intent)
        }

        /**
         * Stops this service.
         *
         * @param context the application context.
         */
        fun stop(context: Context) {
            context.stopService(Intent(context, DownloadService::class.java))
        }

        /**
         * Returns the status of the service.
         *
         * @param context the application context.
         * @return true if the service is running, false otherwise.
         */
        fun isRunning(context: Context): Boolean {
            return context.isServiceRunning(DownloadService::class.java)
        }
    }

    private val downloadManager: DownloadManager by injectLazy()
    private val downloadPreferences: DownloadPreferences by injectLazy()

    /**
     * Wake lock to prevent the device to enter sleep mode.
     */
    private lateinit var wakeLock: PowerManager.WakeLock

    private lateinit var scope: CoroutineScope

    override fun onCreate() {
        scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
        startForeground(Notifications.ID_DOWNLOAD_CHAPTER_PROGRESS, getPlaceholderNotification())
        wakeLock = acquireWakeLock(javaClass.name)
        _isRunning.value = true
        listenNetworkChanges()
    }

    override fun onDestroy() {
        scope.cancel()
        _isRunning.value = false
        downloadManager.downloaderStop()
        if (wakeLock.isHeld) {
            wakeLock.release()
        }
    }

    // Not used
    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        return START_NOT_STICKY
    }

    // Not used
    override fun onBind(intent: Intent): IBinder? {
        return null
    }

    private fun downloaderStop(@StringRes string: Int) {
        downloadManager.downloaderStop(getString(string))
    }

    private fun listenNetworkChanges() {
        ReactiveNetwork()
            .observeNetworkConnectivity(applicationContext)
            .onEach {
                withUIContext {
                    if (isOnline()) {
                        if (downloadPreferences.downloadOnlyOverWifi().get() && !isConnectedToWifi()) {
                            downloaderStop(R.string.download_notifier_text_only_wifi)
                        } else {
                            val started = downloadManager.downloaderStart()
                            if (!started) stopSelf()
                        }
                    } else {
                        downloaderStop(R.string.download_notifier_no_network)
                    }
                }
            }
            .catch { error ->
                withUIContext {
                    logcat(LogPriority.ERROR, error)
                    toast(R.string.download_queue_error)
                    stopSelf()
                }
            }
            .launchIn(scope)
    }

    private fun getPlaceholderNotification(): Notification {
        return notificationBuilder(Notifications.CHANNEL_DOWNLOADER_PROGRESS) {
            setContentTitle(getString(R.string.download_notifier_downloader_title))
        }.build()
    }
}
